### PR TITLE
Food Ownership Changes

### PR DESCRIPTION
**Pull request by bunniewormy**
## Status

(pick one)

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.

## Description

Food ownership now works similiarly to mines - the "owner" of a food item is the person who bought it or first picked it up. You also become owner of food items taken from enemies. When somebody in your team picks up your food items, you are still the owner. The point of this PR is that the coins for healing people are given to the person who buys the food, rather than the person that throws food at damaged teammates.

Moreover, you can also see owner of a food item when you hover your mouse over a food item. This works the same as my recent mine PR https://github.com/transhumandesign/kag-base/pull/894

### CODE REVIEW

File: Entities/Items/Food/Eatable.as [hunk 1]
```diff
@@ -78,20 +97,43 @@ void onCollision(CBlob@ this, CBlob@ blob, bool solid)
...
+        this.Sync("healer", true);
+    }
+}
 
-	this.set_u16("healer", p.getNetworkID());
+void onThisAddToInventory(CBlob@ this, CBlob@ inventoryBlob)
+{
+    CPlayer@ player = inventoryBlob.getPlayer();
+
+    if (player !is null && (this.getTeamNum() != inventoryBlob.getTeamNum() || !this.exists("healer")))
```
Comment (AsuMagic) [inline Entities/Items/Food/Eatable.as:127]: That `if` condition logic is duplicated with the `onAttach` function, make a function for this, e.g. `void setHealer(CBlob@ food_blob, CPlayer@ player)` so you can call it like `setHealer(@this, @player);`

File: Entities/Items/Food/Eatable.as [hunk 2]
```diff
@@ -10,6 +10,25 @@ void onInit(CBlob@ this)
 	this.addCommandID(heal_id);
 }
 
+void onTick(CBlob@ this)
```
Comment (AsuMagic) [inline Entities/Items/Food/Eatable.as:13]: The whole hook seems unnecessary as you could get the username in the `onRender` hook instead, without having to use a separate property for the player name and network ID.

File: Entities/Items/Food/Eatable.as [hunk 3]
```diff
@@ -78,20 +97,43 @@ void onCollision(CBlob@ this, CBlob@ blob, bool solid)
...
+
+    if (player !is null && (this.getTeamNum() != inventoryBlob.getTeamNum() || !this.exists("healer")))
+    {
+        if (isServer())
+        {
+            this.server_setTeamNum(inventoryBlob.getTeamNum());
+        }
+
+        this.set_u16("healer", player.getNetworkID());
+        this.Sync("healer", true);
```
Comment (AsuMagic) [inline Entities/Items/Food/Eatable.as:135]: I don't think the `Sync` is necessary, as I believe this hook should get called both on client and server side already. Should check whether removing that line breaks the behavior on dedicated servers.

File: Entities/Items/Food/Food.as [hunk 4]
```diff
@@ -15,7 +15,26 @@ void onInit(CBlob@ this)
 
 	this.getSprite().SetFrameIndex(index);
 	this.SetInventoryIcon(this.getSprite().getConsts().filename, index, Vec2f(16, 16));
-	this.server_setTeamNum(0); // blue fishy like in sprite sheet
+	// this.server_setTeamNum(0); // blue fishy like in sprite sheet
```
Comment (AsuMagic) [inline Entities/Items/Food/Food.as:18]: Not sure why this is commented out now, should probably clarify why in a comment if intended

**Review by AsuMagic — CHANGES_REQUESTED** [review]

File: Entities/Items/Food/Eatable.as (hunk 2 above)
Comment (bunniewormy) [reply Entities/Items/Food/Eatable.as]: should be fine now

File: Entities/Items/Food/Eatable.as (hunk 1 above)
Comment (bunniewormy) [reply Entities/Items/Food/Eatable.as]: added setHealer to EatCommon.as

File: Entities/Items/Food/Eatable.as (hunk 3 above)
Comment (bunniewormy) [reply Entities/Items/Food/Eatable.as]: should be fine now

File: Entities/Items/Food/Food.as [hunk 5]
```diff
@@ -15,7 +15,44 @@ void onInit(CBlob@ this)
...
+	{
+		u16 healerID = blob.get_u16("healer");
+		CPlayer@ healer = getPlayerByNetworkId(healerID);
+		if(healer !is null)
+		{
+			healerName = healer.getUsername(); 
+		}
+	}
+
+	blob.Sync("healer", true);
```
Comment (AsuMagic) [inline Entities/Items/Food/Food.as:49]: That should not be here

Comment (bunniewormy) [reply Entities/Items/Food/Food.as]: fixed

**Review by AsuMagic — DISMISSED** [review]
LGTM

**Pull request closed without merging**